flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerCleanupITCase.java

```diff
@@ -0,0 +1,261 @@
+						int blobPort = (Integer) Await.result(future, remaining());
+
+						// upload a blob
+						BlobKey key1;
+						bc = new BlobClient(new InetSocketAddress("localhost", blobPort), config);
+						try {
+							key1 = bc.put(jid, new byte[10]);
+						} finally {
+							bc.close();
+						}
+						jobGraph.addBlob(key1);
+
+						// Submit the job and wait for all vertices to be running
+						jobManagerGateway.tell(
+							new JobManagerMessages.SubmitJob(
+								jobGraph,
+								ListeningBehaviour.EXECUTION_RESULT),
+							testActorGateway);
+						expectMsgClass(JobManagerMessages.JobSubmitSuccess.class);
+
+
+						if (finalState == ExecutionState.FAILED) {
+							// fail a task so that the job is going to be recovered (we actually do not
+							// need the blocking part of the invokable and can start throwing right away)
+							FailingBlockingInvokable.unblock();
+
+							// job will get restarted, BlobCache may re-download the BLOB if already deleted
+							// then the tasks will fail again and the restart strategy will finalise the job
+
+							expectMsgClass(JobManagerMessages.JobResultFailure.class);
+						} else if (finalState == ExecutionState.CANCELED) {
+							jobManagerGateway.tell(
+								new JobManagerMessages.CancelJob(jid),
+								testActorGateway);
+							expectMsgClass(JobManagerMessages.CancellationResponse.class);
+
+							// job will be cancelled and everything should be cleaned up
+
+							expectMsgClass(JobManagerMessages.JobResultFailure.class);
+						} else {
+							expectMsgClass(JobManagerMessages.JobResultSuccess.class);
+						}
+
+						// both BlobServer and BlobCache should eventually delete all files
+
+						File[] blobDirs = blobBaseDir.listFiles(new FilenameFilter() {
+							@Override
+							public boolean accept(File dir, String name) {
+								return dir.getName().startsWith("blobStore-");
+							}
+						});
+						assertNotNull(blobDirs);
+						for (File blobDir : blobDirs) {
+							waitForEmptyBlobDir(blobDir);
+						}
+
+					} catch (Exception e) {
+						e.printStackTrace();
+						fail(e.getMessage());
+					} finally {
+						if (bc != null) {
+							try {
+								bc.close();
+							} catch (IOException ignored) {
+							}
+						}
+						if (cluster != null) {
+							cluster.shutdown();
+						}
+					}
+				}
+			};
+		}};
+
+		// after everything has been shut down, the storage directory itself should be empty
+		assertArrayEquals(new File[] {}, blobBaseDir.listFiles());
+	}
+
+	/**
+	 * Waits until the given {@link org.apache.flink.runtime.blob.BlobService} storage directory
+	 * does not contain any job-related folders any more.
+	 *
+	 * @param blobDir
+	 * 		directory of a {@link org.apache.flink.runtime.blob.BlobServer} or {@link
+	 * 		org.apache.flink.runtime.blob.BlobCache}
+	 *
+	 * @see org.apache.flink.runtime.blob.BlobUtils
+	 */
+	private static void waitForEmptyBlobDir(File blobDir) {
+		while (true) {
```

Comment:
Shall we make a small pause in between directory list calls to not burn too many cycles?